### ISSUE TITLE
Get the operation DofTransformation->InvTransformPrimal as a dense matrix

### ISSUE THREAD

**Issue by adam-sim-dev** [mfem/mfem#3568]
Hi all,

I am wondering if I can get the function `DofTransformation->InvTransformPrimal` as an operator (dense matrix), let's say `transp`,

Now mfem commonly use
```
Vector loc_data;
if(doftrans)
{
  doftrans->InvTransformPrimal(loc_data)
}
```
I want to use
`DenseMatrix transp;`
`DofTransformation->Get_InvTransformPrimal_Operator(transp);`
`transp.MultTranspose(loc_data, loc_data_new)`
loc_data_new is now equivalent to loc_data.

Best,
Adam

**Comment by mlstowell** [comment]
Hi, @adam-sim-dev,

We certainly could add such a method but it does not currently exist. The reason is that `transp` would be mostly zeros so filling and applying such an operator would not be as efficient as calling `doftrans->InvTransformPrimal(loc_data)` or `doftrans->TransformDual(loc_data)` (I believe `transp.MultTranspose` would produce the same result as `doftrans->TransformDual(loc_data)`). For example, on a 2nd order tetrahedron there are 20 degrees of freedom, 12 of which are on edges and therefore not altered by the `DofTransformation`. In this case the `transp` matrix would be a 20x20 dense matrix with twelve ones along the diagonal and four 2x2 diagonal blocks for the 4 pairs of DoFs internal to the four faces. This would produce a dense matrix with potentially only 28 non-zero values i.e. 93% empty.

That being said, I do recognize that a `DenseMatrix` has certain advantages. If this is something you need I would suggest that you consider working with the 2x2 submatrices. This is how the implementations are done internally. We don't currently have an interface that we could expose to the user but we could add this. Such an interface would need to provide the appropriate offsets into the local vectors, references to the `DenseMatrix` objects, as well as the numbers and sizes of the blocks (currently 2x2 blocks handle the Nedelec case but other requirements might arise in the future). Does this sound like something you would find useful?

Best wishes,
Mark

**Comment by adam-sim-dev** [comment]
Hi @mlstowell ,

Thanks for the explanation. I also see your comments https://github.com/mfem/mfem/pull/3379#issuecomment-1362028238

This feature is useful for me because for my application I have to call doftrans->InvTransformPrimal thousands of times. If I could save this operator as a DenseMatrix, then I can just do linear algebra and will save a lot of time.

Best,
Adam

**Comment by mlstowell** [comment]
Hi, @adam-sim-dev,

`doftrans->InvTransformPrimal` has the potential to produce 256 different 20x20 operators based on the orientations of the four faces of a tetrahedron (assuming 2nd order basis functions). On the other hand there are only six of the 2x2 matrices (and only four of them appear under usual circumstances).

The `DofTransformation` is a relatively new class and I'd be happy to work with you to find more efficient ways to provide this information in your use-case but I'd like to better understand your requirements first.

Best wishes,
Mark

**Comment by adam-sim-dev** [comment]
Hi, @mlstowell 

What I do is very similar to GridFunction::GetVectorValue(), i.e., the electromagnetic field interpolation.

I want to
1) Save the shape function as a DenseMatrix shape , then e = shape * sol.
2) For the high-order FEs, I need the InvTransformPrimal (in the matrix form transform), then e = shape * transform * sol.

I am solving time-varying EM problems, so I want to save the operator interpolation = shape * transform then I can calculate the electric field quickly without the tedious operation in GridFunction::GetVectorValue()

**Comment by adam-sim-dev** [comment]
> Hi, @adam-sim-dev,
> 
> `doftrans->InvTransformPrimal` has the potential to produce 256 different 20x20 operators based on the orientations of the four faces of a tetrahedron (assuming 2nd order basis functions). On the other hand there are only six of the 2x2 matrices (and only four of them appear under usual circumstances).
> 
For one element, if given the element number and the IntegrationPoint, is `doftrans->InvTransformPrimal` unique?

**Comment by mlstowell** [comment]
Yes, it would be. It depends only on the topology of the mesh so even the integration point is not relevant. It will vary with element number but that's all.

**Comment by mlstowell** [comment]
Okay, this makes sense now. One way you could do this with the existing interface is to transform your `shape` matrix.  The correct call should be `doftrans->TransformDualRows(shape);`. This replaces the `DenseMatrix shape` with the product `shape * transform`. If you need to recompute `shape` at multiple integration points you would need to call `doftrans->TransformDualRows(shape);` for each `shape` matrix but this should be more efficient than performing a matrix-matrix multiply.

**Comment by adam-sim-dev** [comment]
> Okay, this makes sense now. One way you could do this with the existing interface is to transform your `shape` matrix.  The correct call should be `doftrans->TransformDualRows(shape);`. This replaces the `DenseMatrix shape` with the product `shape * transform`. If you need to recompute `shape` at multiple integration points you would need to call `doftrans->TransformDualRows(shape);` for each `shape` matrix but this should be more efficient than performing a matrix-matrix multiply.

Oh, thank you very much. I have not noticed the function TransformDualRows(shape) yet. I will have a try.

**Comment by adam-sim-dev** [comment]
> Okay, this makes sense now. One way you could do this with the existing interface is to transform your `shape` matrix. The correct call should be `doftrans->TransformDualRows(shape);`. This replaces the `DenseMatrix shape` with the product `shape * transform`. If you need to recompute `shape` at multiple integration points you would need to call `doftrans->TransformDualRows(shape);` for each `shape` matrix but this should be more efficient than performing a matrix-matrix multiply.

Hi @mlstowell , I have a similar application case for the H1 space. `scalar = shape * sol`, while sol is in H1. Now, `shape` is a vector, not a DenseMatrix. Which function should I call to transform the `shape` vector? similar to https://github.com/mfem/mfem/blob/5e32ae87b48a7da7d92621c6d152a13497b69d57/fem/gridfunc.cpp#L465-L468
